### ISSUE TITLE
cannot locate symbol "_ZNSt6__ndk14coutE" with c++_shared

### ISSUE THREAD

**Issue by jsren** [android-ndk/ndk#173]
Running the application fails with `CANNOT LINK EXECUTABLE: cannot locate symbol "_ZNSt6__ndk14coutE"` when built with Clang or GCC (4.9) & *c++_shared*. Works fine using *c++_static*. (android-ndk-r12b, Windows.)

To reproduce:

```
NDK_TOOLCHAIN_VERSION = clang # or 4.9
APP_PLATFORM := android-24
APP_STL      := c++_shared
APP_ABI      := armeabi-v7a
APP_BUILD_SCRIPT := Android.mk
# ------------------------------------------
LOCAL_PATH := $(call my-dir)
include $(CLEAR_VARS)
LOCAL_MODULE := test
LOCAL_SRC_FILES := test.cpp
include $(BUILD_EXECUTABLE)
# ------------------------------------------
#include <iostream>
#include <string>
int main()
{
	std::cout << "Example\n";
	return 0;
}
# ------------------------------------------
> /system/bin/test 
CANNOT LINK EXECUTABLE: cannot locate symbol "_ZNSt6__ndk14coutE" referenced by "/system/bin/test"...
```

**Comment by DanAlbert** [comment]
`adb shell LD_LIBRARY_PATH=/system/bin /system/bin/test` (assuming you pushed `libc++_shared.so` to `/system/bin` as well)

**Comment by jsren** [comment]
Actually I definitely pushed `libc++_shared.so` to `/system/lib`. It's not in `bin`, and when I add `LD_LIBRARY_PATH=/system/lib` it works. This behaviour seems to be different from the previous version we used (r11).

**Comment by jsren** [comment]
Also with r11 a load error gives explicitly ```CANNOT LINK EXECUTABLE: library "libc++_shared.so" not found```

**Comment by DanAlbert** [comment]
> APP_PLATFORM := android-24

Possibly related to https://developer.android.com/preview/behavior-changes.html#ndk ?

By putting the library in `/system/lib` you've turned it into a system library. I don't actually know how a binary gets identified as being a non-system library. @dimitry- can shed some light on that though.

If its based on `APP_ABI` then that would explain the difference between r11 and r12 too; r11 didn't have android-24.

**Comment by dimitry-** [comment]
it should not affect executables, this mechanism works only for app libraries.

is there a difference between readelf -d <binary> compiled with r11 and r12?

in other words is libc++_shared.so on the list for r12?

**Comment by DanAlbert** [comment]
Works for me. What were you testing with? My test was an internal build of N, so I suppose this might be something fixed between your build and now.

**Comment by jsren** [comment]
I've tried again and have been unable to reproduce. There have been a number of code & environment changes since then, however. What was strange was that the issue was present on two independent machines. I suspect that it might have something to do with Cygwin - IIRC there's a difference in path handling since the previous versions? 

When I updated the NDK and switched to Clang I originally had issues where it was somehow launching the Cygwin `as`, so I suspect there was something going on with paths there.
